VideoPress block: Initial setting tests

SiobhyB opened the pull request:
Part of the coverage efforts outlined in https://github.com/wordpress-mobile/gutenberg-mobile/issues/5734

## Description

The changes in this PR introduce tests that ensure the VideoPress block's title, description, and Playback settings can be successfully updated, as well as some of the settings in the Playback Bar Color and Privacy and Rating panels.

_Note, the following are not covered by this PR, and will be worked on in separate PRs:_

* Rating and Privacy picker selection.
* Various color selections for Playback Bar Color.
* Setting synchronization between the app and the web. 
 
## To test:

* Verify the HTML in [the snapshot file](https://github.com/wordpress-mobile/gutenberg-mobile/blob/f8736c8dd5a706deb525308e8a5e637e6804a331/src/test/videopress/__snapshots__/edit.js.snap) matches with the reality of what the block's HTML would look after making the relevant changes.
* Run the `TEST_RN_PLATFORM=ios npm run test src/test/videopress/edit.js` and `TEST_RN_PLATFORM=android npm run test src/test/videopress/edit.js` commands from the terminal to verify the tests pass as expected. Also, verify the tests on this PR pass.

<hr />

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

SiobhyB committed:
feat: add constants for VideoPress block settings

This commit introduces a set of constants related to the VideoPress block settings, including default props, block HTML, playback settings, color settings, rating options, privacy options, and additional settings. These constants will be used to manage and manipulate VideoPress block properties in a more organised and maintainable manner.

SiobhyB committed:
tests: Groundwork to open block settings

This commit contains the code necessary to initialize the editor with existing HTML and open the block's settings. It serves as the groundwork for future tests, which will actually change the block's settings.

SiobhyB committed:
test: Update title setting

SiobhyB committed:
test: Update description setting

SiobhyB committed:
test: Loop through and toggle playback settings

SiobhyB committed:
test: Loop through and update playback bar color

SiobhyB committed:
test: Loop through and select rating options

SiobhyB committed:
test: Loop through and select privacy options

SiobhyB committed:
test: Loop through remaining privacy options

SiobhyB committed:
add: Snapshots for all setting tests

SiobhyB committed:
refactor: Update Gutenberg reference

SiobhyB commented on src/test/videopress/edit.js:
```diff
@@ -44,5 +55,116 @@ describe( 'VideoPress block', () => {
 		expect( getEditorHtml() ).toBe( expectedHtml );
+
+		// Reset editor to avoid side effects from built-in timers in addBlock() function on iOS
+		await initializeEditor();
```
Without this reset, the first test in the second test suite (and only the first test) was consistently failing (no matter what that test was). I believe this may be the result of the extra delay added to the `addBlock` helper in https://github.com/WordPress/gutenberg/pull/50381.

fluiddot replied:
@SiobhyB what error are you getting? Asking in case it's related to this issue https://github.com/wordpress-mobile/gutenberg-mobile/issues/5742.

In theory, we wouldn't need to initialize the editor after the test finishes 🤔 .

SiobhyB replied:
@fluiddot, if you comment out this line and run `TEST_RN_PLATFORM=ios npm run test src/test/videopress/edit.js`, the first test in the second test suite (no matter what that test is) fails with the following:

```
    TypeError: Cannot read property 'presentPicker' of null

      195 | 			if ( autoOpen && isIOS ) {
      196 | 				this.pickerTimeout = setTimeout(
    > 197 | 					() => this.picker.presentPicker(),
          | 					                  ^
      198 | 					PICKER_OPENING_DELAY
      199 | 				);
      200 | 			} else {
```

If you navigate to that code directly and change the conditional so that `this.picker.presentPicker` runs without a delay, the test passes. Similarly, the test will also pass after commenting out the `addBlock` call, which makes me think that the failure is related to timing/delays, leading to this hack.

The PR related to the issue you linked to, https://github.com/WordPress/gutenberg/commit/023631c27cfecd9684d85a4bf92570158660edbf, doesn't resolve the issue unfortunately.

fluiddot replied:
Ah ok, seems we'd need to run the timers after adding the block via the inserter menu to avoid this. I tried adding the following change and seemed to work.

```patch
diff --git forkSrcPrefix/src/test/videopress/edit.js forkDstPrefix/src/test/videopress/edit.js
index 7dec3ce42eff2e6b1af50e7cffe8eda9a32acf9a..94c975a65299c23d146f2109c77ae71247a38a8d 100644
--- forkSrcPrefix/src/test/videopress/edit.js
+++ forkDstPrefix/src/test/videopress/edit.js
@@ -45,6 +45,8 @@ describe( 'VideoPress block', () => {
 		// Add block
 		await addBlock( screen, 'VideoPress' );
 
+		jest.runOnlyPendingTimers();
+
 		// Get block
 		const videoPressBlock = await getBlock( screen, 'VideoPress' );
 		expect( videoPressBlock ).toBeVisible();
```

I was curious why I didn't experience this in the upload test cases. However, in that case, since the media picker is opened during the editor initialization, the timers are automatically run.

SiobhyB replied:
Ah, wow, thank you Carlos, such a seemingly simple solution to something I was racking my brain over. 🤦‍♀️ I've gone ahead to apply your fix in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751/commits/d56e303a6ab9570e5acd897aec9f06c8725af4c7.

SiobhyB commented on src/test/videopress/constants.js:
```diff
@@ -0,0 +1,27 @@
+	'Preload Metadata',
+];
+
+export const PLAYBACK_BAR_COLOR_SETTINGS = [ 'Dynamic color' ];
```
I plan to work on adding the color selections in a future PR.

SiobhyB committed:
refactor: Address linting errors

SiobhyB committed:
refactor: Update Gutenberg reference

SiobhyB committed:
refactor: Remove unnecessary "await"

SiobhyB committed:
refactor: Rearrange helper files under new folder

SiobhyB committed:
fix: Exclude test files under "local-helpers"

SiobhyB committed:
refactor: Align JSDocs

SiobhyB committed:
refactor: Align JSDocs comment

SiobhyB committed:
refactor: Remove privacy and rating settings

SiobhyB committed:
refactor: Update Gutenberg submodule reference

SiobhyB committed:
refactor: Remove unused import

SiobhyB commented on src/test/videopress/local-helpers/constants.js:
```diff
@@ -0,0 +1,23 @@
+	'Preload Metadata',
+];
+
+export const PLAYBACK_BAR_COLOR_SETTINGS = [ 'Dynamic color' ];
```
This will be added to in an upcoming PR.

fluiddot commented on src/test/videopress/local-helpers/constants.js:
Ah interesting, this will be also helpful for the upload test cases.

fluiddot commented on src/test/videopress/local-helpers/utils.js:
```diff
@@ -0,0 +1,50 @@
+ */
+import { VIDEOPRESS_BLOCK_HTML } from './constants';
+
+export const initializeBlockWithHTML = async (
+	initialHtml = VIDEOPRESS_BLOCK_HTML
+) => {
+	const screen = await initializeEditor( { initialHtml } );
+
+	return { ...screen };
+};
```
Not sure if this function is reducing the test boilerplate 🤔 . It would be the same as doing:
```
const screen = await initializeEditor( { initialHtml: VIDEOPRESS_BLOCK_HTML  } );
```
vs
```
const screen = await initializeBlockWithHTML( VIDEOPRESS_BLOCK_HTML );
```
Do we expect to add more functionality in the future?

SiobhyB replied:
Ah, you're right, thanks for pointing this out. I don't expect to add further functionality here in the future, I believe it ended up this way due to some local explorations and changes. I've gone ahead to remove this as a helper in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751/commits/5293b8b128b97f0df329446a693c9c8c48ae4dea.

fluiddot commented on src/test/videopress/edit.js:
```diff
@@ -44,5 +51,84 @@ describe( 'VideoPress block', () => {
+	 * PLAYBACK SETTINGS
+	 * Loop through each of the playback settings and toggle on/off
+	 */
+	PLAYBACK_SETTINGS.forEach( ( setting ) => {
+		it( `should update Playback Settings section's ${ setting } setting`, async () => {
```
Jest provides [a function](https://jestjs.io/docs/api#testeachtablename-fn-timeout) to run the same tests with different data, sharing here in case you want to use it.

fluiddot commented on src/test/videopress/edit.js:
```diff
@@ -44,5 +51,84 @@ describe( 'VideoPress block', () => {
+} );
+
+describe( "Update VideoPress block's settings", () => {
+	let screen;
+
+	beforeEach( async () => {
+		// Arrange
+		screen = await initializeBlockWithHTML();
+
+		await selectAndOpenBlockSettings( screen );
+	} );
```
Nitpick: Usually in integration tests, we have all the arrange, act, and expectation parts contained within the `it` function. This introduces a bit of redundancy but in general, helps readability as you can see in the same place what the test case is doing.

SiobhyB replied:
Happy to go with the standard here, updated in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751/commits/77aa2c9e8a0dd73ac4dca1bb0c7b7731c96eec44.

fluiddot reviewed (commented):
When running the tests I got several warnings with the following message
```
Warning: An update to Player inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
```

Seems the player component is being updated after executing some of the actions. It would be great to incorporate a check for this in the test cases to avoid the warning.

SiobhyB committed:
refactor: Wrap state change in act

This addresses the feedback received here: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751#pullrequestreview-1420887276

SiobhyB committed:
refactor: Remove unnecessary helper

Refactored following feedback here: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751#discussion_r1190074663

SiobhyB committed:
refactor: Move all 'act' logic is to it statement

As per this feedback, this introduces redundancy, but is a tradeoff for readability: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751#discussion_r1190088716

SiobhyB committed:
fix: Run timers after adding block

Discussion here: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751#discussion_r1189549726

SiobhyB committed:
Merge branch 'trunk' into add/videopress-setting-tests

SiobhyB commented:
@fluiddot thank you for the review 🙇‍♀️ 

> Seems the player component is being updated after executing some of the actions. It would be great to incorporate a check for this in the test cases to avoid the warning.

I've gone ahead to address this in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751/commits/881272b3af89171b610adaa9d32d3117bee6a86e, and believe I've addressed your other feedback too. Let me know if there's anything else that needs improving here.

fluiddot commented on src/test/videopress/edit.js:
```diff
@@ -38,6 +45,8 @@ describe( 'VideoPress block', () => {
 		// Add block
 		await addBlock( screen, 'VideoPress' );
 
+		jest.runOnlyPendingTimers();
+
```
When running the command `TEST_RN_PLATFORM=ios npm run test src/test/videopress/edit.js` I'm getting the following warning:
```
 A function to advance timers was called but the timers APIs are not mocked with fake timers. Call `jest.useFakeTimers({legacyFakeTimers: true})` in this test file or enable fake timers for all tests by setting {'enableGlobally': true, 'legacyFakeTimers': true} in Jest configuration file.
    Stack Trace:

          46 | 		await addBlock( screen, 'VideoPress' );
          47 |
        > 48 | 		jest.runOnlyPendingTimers();
             | 		     ^
          49 |
          50 | 		// Get block
          51 | 		const videoPressBlock = await getBlock( screen, 'VideoPress' );

          Error:
          at FakeTimers._checkFakeTimers (node_modules/@jest/fake-timers/build/legacyFakeTimers.js:394:13)
          at FakeTimers.runOnlyPendingTimers (node_modules/@jest/fake-timers/build/legacyFakeTimers.js:189:10)
          at Object.runOnlyPendingTimers (src/test/videopress/edit.js:48:8)
          at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
          at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
          at tryCallOne (gutenberg/node_modules/promise/lib/core.js:37:12)
          at gutenberg/node_modules/promise/lib/core.js:123:15
          at flush (gutenberg/node_modules/asap/raw.js:50:29)
```

Probably we could use the `withFakeTimers` helper to address the issue:
```
// When the block is inserted, it automatically opens the media picker.
// On iOS, this picker is displayed using a timer, so we need to run it
// to allow any DOM update.
await withFakeTimers( () => jest.runOnlyPendingTimers() );
```

SiobhyB replied:
Thank you, I've gone ahead to implement that in https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751/commits/b0be14a9e46dcbff0aa1bfd02d910a67f7e8d615.

fluiddot reviewed (commented):
I noticed several logs with the following message when running the tests. Seems the block is trying to fetch the metadata via `fetchVideoItem` function but fails. If we don't need that data for these tests, maybe we could mock that function, WDYT?

```
videopress:lib:fetch-video-item updating retry from 0 to 1
```

SiobhyB committed:
Merge branch 'trunk' into add/videopress-setting-tests

SiobhyB committed:
fix: Ensure timer runs as expected after addBlock

Further context: https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751#discussion_r1190821054

SiobhyB committed:
refactor: Remove unused import

fluiddot reviewed (approved):
LGTM 🎊 !

I confirmed that all tests passed ✅ . Not a blocker, but we might consider adding the mock I suggested [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5751#pullrequestreview-1422037209) to avoid the fetch failures when the block tries to sync the metadata.

SiobhyB commented:
Thanks @fluiddot! I'll go ahead to merge as-is and address mocking `fetchVideoItem` as part of a separate PR. 🙇‍♀️
